### PR TITLE
Refactor Controller API and Adding the missing APIs

### PR DESCRIPTION
**Pull request by ShanBai6**
1. Add abstract class for methods shared across controller API and refactor the existing ones.
2. Add the remaining controller API and refactor the agent utility classes as needed.

### CODE REVIEW

**Comment by jirassimok** [comment]
One specific change you can make:

ApiRequester can not accept null query parameters; there are two steps that should be taken about this:
First, each query parameter should only be added if given by the caller. To facilitate this, this method should be added to AbstractControllerAPI, and used for all of the query parameters that need to be send to the agents:
```java
protected void addQueryParam(ApiRequester request, String name, Object value) {
    if (value != null) {
        request.queryParam(name, value);
    }
}
```

Second, `ApiRequester#queryParam(String, Object...)` should start with this block:
```java
for (Object o : values) {
    requireNonNull(o);
}
```

**Comment by jirassimok** [comment]
The DateParameter stuff should probably be noted more prominently, probably in their own commit, but that would be a bit tricky.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoRpmController.java [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParser;
 import com.google.inject.Inject;
+import com.teradata.prestomanager.common.StopType;
```
Comment (cawallin) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoRpmController.java]: Unnecessary change?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/ControlAPI.java [hunk 2]
```diff
@@ -15,7 +15,8 @@
 
 import com.google.inject.Inject;
 import com.teradata.prestomanager.agent.PrestoRpmController;
-import com.teradata.prestomanager.agent.StopType;
+import com.teradata.prestomanager.common.StopType;
+import io.airlift.log.Logger;
```
Comment (cawallin) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/ControlAPI.java]: Unnecessary change?

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/LogsAPI.java [hunk 3]
```diff
@@ -45,20 +40,6 @@
         date = "2017-06-23T09:53:13.549-04:00")
 public class LogsAPI
 {
-    private static final String DEFAULT_DATE = "DEFAULT";
```
Comment (cawallin) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/LogsAPI.java]: Ideally, fixing DateParameter should be a separate commit.

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/JaxrsParameter.java [hunk 4]
```diff
@@ -39,6 +39,8 @@ public JaxrsParameter(String s)
     protected abstract T parseString(String s)
             throws ParseException;
 
+    protected abstract String getString();
```
Comment (cawallin) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/JaxrsParameter.java]: Why getString() and not toString()?

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/StopType.java [hunk 5]
```diff
@@ -11,7 +11,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.teradata.prestomanager.agent;
```
Comment (cawallin) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/StopType.java]: Ideally, this also would be in a separate commit

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/JaxrsParameter.java (hunk 4 above)
Comment (cawallin) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/JaxrsParameter.java]: Instead of getString(), consider overriding toString().

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            RequestDispatcher requestDispatcher)
+    {
+        super(forwardingClient, requestDispatcher);
+    }
+
+    @POST
+    @Path("/start")
+    @Produces(MediaType.TEXT_PLAIN)
+    public Response startPresto(
+            @QueryParam("scope") String scope,
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: How does this handle if the scope and/or nodeId list aren't specified?

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    public Response stopPresto(
+            @QueryParam("stopType") @DefaultValue("GRACEFUL") StopType stopType,
+            @QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<UUID> nodeId)
+    {
+        ApiRequester.Builder apiRequester = requesterBuilder(ControllerControlAPI.class)
+                .httpMethod(POST)
+                .pathMethod("stopPresto");
+
+        OptionalQueryParam(apiRequester, "stopType", stopType);
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: If this has a default value, it's not optional, right?

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        return forwardRequest(scope, apiRequester.build(), nodeId);
+    }
+
+    @POST
+    @Path("/restart")
+    public Response restartPresto(
+            @QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<UUID> nodeId)
+    {
+        ApiRequester apiRequester = requesterBuilder(ControllerControlAPI.class)
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: What type of stop does restart do? It seems like (except for the coordinator) you could choose to have a graceful stop here too.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            @QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<UUID> nodeId)
+    {
+        ApiRequester.Builder apiRequester = requesterBuilder(ControllerLogsAPI.class)
+                .httpMethod(GET)
+                .accept(MediaType.TEXT_PLAIN)
+                .pathMethod("getLog")
+                .resolveTemplate("file", file);
+
+        OptionalQueryParam(apiRequester, "from", fromDate.getString());
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java]: from, to, and level don't seem optional

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            @QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<UUID> nodeId)
+    {
+        ApiRequester.Builder apiRequester = requesterBuilder(ControllerLogsAPI.class)
+                .httpMethod(DELETE)
+                .accept(MediaType.TEXT_PLAIN)
+                .pathMethod("deleteLog")
+                .resolveTemplate("file", file);
+
+        OptionalQueryParam(apiRequester, "to", toDate.getString());
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java]: ditto on optional

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/AbstractControllerAPI.java [hunk 11]
```diff
@@ -34,6 +34,13 @@ protected AbstractControllerAPI(
         this.requestDispatcher = requestDispatcher;
     }
 
+    protected void OptionalQueryParam(ApiRequester.Builder request, String name, Object value)
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/AbstractControllerAPI.java]: This method needs to start with lower case

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerPackageAPI.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    @Produces({MediaType.TEXT_PLAIN})
+    public Response install(String urlToFetchPackage,
+            @QueryParam("checkDependencies") @DefaultValue("true") boolean checkDependencies,
+            @QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<UUID> nodeId)
+    {
+        ApiRequester.Builder apiRequester = requesterBuilder(ControllerPackageAPI.class)
+                .httpMethod(PUT)
+                .accept(MediaType.TEXT_PLAIN)
+                .entity(Entity.entity(urlToFetchPackage, MediaType.TEXT_PLAIN));
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerPackageAPI.java]: If urlToFetchPackage can't be null, you should add an explicit null check

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            @QueryParam("level") @DefaultValue("ALL") String level,
+            @QueryParam("n") Integer maxEntries,
+            @QueryParam("scope") String scope,
+            @QueryParam("nodeId") List<UUID> nodeId)
+    {
+        ApiRequester.Builder apiRequester = requesterBuilder(ControllerLogsAPI.class)
+                .httpMethod(GET)
+                .accept(MediaType.TEXT_PLAIN)
+                .pathMethod("getLog")
+                .resolveTemplate("file", file);
```
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java]: Make sure to do a null check for file, it needs to be non-null.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoRpmController.java (hunk 1 above)
Comment (ShanBai6) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoRpmController.java]: Since `Controller` already have dependency on `common` module, we move that from agent to common so that we could use it in `Controller` module

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/JaxrsParameter.java (hunk 4 above)
Comment (jirassimok) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/JaxrsParameter.java]: `null` is a poor string representation.

It might make more sense to return an `Optional<String>`. Then it might also make sense to change the behavior of `get()` to behave like `Optional::get`.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java (hunk 6 above)
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: It is handled when it calls `forwardRequest` inside `RequestDispatcher`

File: presto-manager-common/src/main/java/com/teradata/prestomanager/common/StopType.java (hunk 5 above)
Comment (ShanBai6) [inline presto-manager-common/src/main/java/com/teradata/prestomanager/common/StopType.java]: Yes, will do that next time encountering this situation. Already have too many other ones on top of it.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java (hunk 13 above)
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java]: It is dealt with inside Agent utility

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/AbstractControllerAPI.java (hunk 11 above)
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/AbstractControllerAPI.java]: Updated

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerPackageAPI.java (hunk 12 above)
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerPackageAPI.java]: This is also checked inside the agent. Agent will send back `400 BAD REQUEST` in that situation

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java (hunk 8 above)
Comment (ShanBai6) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: For rpm, restart will always happen no matter of its current state.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java (hunk 9 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerLogsAPI.java]: They are. Unfortunately, Jersey doesn't recognize `Optional` in parameters.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java (hunk 7 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: This is a bit of a tricky situation.

JAX-RS should map values for any parameter type that has a single-String constructor, but Jersey seems to have a bug where `null` is never passed in, so a missing parameter always has a value of `null`.
So to make `DateParameter` not require null checks, I provide a default that is parsed the same way `null` is supposed to be parsed.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java (hunk 8 above)
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: Yes.
For rpm restart Presto is automatically stopped if it is running. 
For upgrade and uninstall I terminate presto.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerPackageAPI.java (hunk 12 above)
Comment (sudarsanm94) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerPackageAPI.java]: I think its better to have a null check here so that `BAD REQUEST` is not forwarded to the Agent.

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java (hunk 7 above)
Comment (jirassimok) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/api/ControllerControlAPI.java]: Resolved:

We replaced `JaxrsParameter` with Jersey's native parameter conversion API, `ParamConverterProvider`.

File: presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoRpmController.java (hunk 1 above)
Comment (cawallin) [inline presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/PrestoRpmController.java]: Ah, was it in the same package before so you didn't have to import it?

File: presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/AbstractControllerAPI.java (hunk 11 above)
Comment (cawallin) [inline presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/AbstractControllerAPI.java]: This isn't updated
